### ISSUE TITLE
SCD30 won't persist ForcedRecalibration reference value

### ISSUE THREAD

**Issue by scottmil** [adafruit/Adafruit_SCD30#13]
- Arduino board:  **INSERT ARDUINO BOARD NAME/TYPE HERE**
- Adafruit ESP32 Huzzah
- Adafruit SCD30

- **INSERT ARDUINOVERSION HERE**
- 1.8.13

- List the steps to reproduce the problem below (if possible attach a sketch or
  copy the sketch code in too): **LIST REPRO STEPS BELOW**
This issue was was originally reported on the Adafruit forum by Jennyfool on April 9, 2021.  

https://forums.adafruit.com/viewtopic.php?f=19&t=177913&p=868236#p868236

I also encountered this problem, and was asked to submit a Github issue by mikeyskylar.

With ASC disabled, I tried multiple times to setForceRecalibrationWithReference to 415 ppm, but the value won't persist when I disconnect power and re-energize the SCD30, but returns to 400 ppm. I have been successful in setting the temperature offset and altitude offsets: (console printout below from modified Adafruit sensor tuning code)

Temperature: 22.58 degrees C
Temperature offset: 2.67 degrees C
Altitude offset: 86 meters
Ambient pressure offset: 0 mBar
Relative Humidity: 46.27 %
Forced Recalibration reference: 400 ppm
Automatic Self Calibration disabled
CO2: 413.355 ppm

My modified Adafruit_SCD30_sensor_tuning.ino
```cpp
// Demo of the available settings for the SCD30
#include <Adafruit_SCD30.h>

Adafruit_SCD30  scd30;

boolean setInterval = false;
int interval = 2;  //originally 2

boolean setPressure = false;
int pressure = 0;

boolean setAltitude = false;
int altitude = 86;  //originally 110

boolean setTempOffset = false;
int tempOffset = 347;

boolean setSelfCalibrationStatus = false; 
boolean enableASC = false;  // false disables ASC

boolean setForceRecalibrationReference = false;
int reference = 415;  //originally 400

void setup(void) {
  Serial.begin(115200);
  while (!Serial) delay(10000);     // will pause until serial console opens

  Serial.println("Adafruit SCD30 Sensor adjustment test!");

  // Try to initialize!
  if (!scd30.begin()) {
    Serial.println("Failed to find SCD30 chip");
    while (1) { delay(10); }
  }
  Serial.println("SCD30 Found!");

  /***
   * The code below will report the current settings for each of the
   * settings that can be changed. To see how they work, uncomment the setting
   * code above a status message and adjust the value
   *
   * **Note:** Since Automatic self calibration and forcing recalibration with
   * a reference value overwrite each other, you should only set one or the other
  ***/


  /*** Adjust the rate at which measurements are taken, from 2-1800 seconds */
  if(setInterval) {
     if (!scd30.setMeasurementInterval(interval)) {
     Serial.println("Failed to set measurement interval");
     while(1){ delay(10);}
     }
  }
  Serial.print("Measurement interval: ");
  Serial.print(scd30.getMeasurementInterval());
  Serial.println(" seconds");


  /*** Restart continuous measurement with a pressure offset from 700 to 1400 millibar.
   * Giving no argument or setting the offset to 0 will disable offset correction
   */
  if(setPressure) {
     if (!scd30.startContinuousMeasurement(pressure)){
        Serial.println("Failed to set ambient pressure offset");
        while(1){ delay(10);}
     }
  }
  Serial.print("Ambient pressure offset: ");
  Serial.print(scd30.getAmbientPressureOffset());
  Serial.println(" mBar");


  /*** Set an altitude offset in meters above sea level.
   * Offset value stored in non-volatile memory of SCD30.
   * Setting an altitude offset will override any pressure offset.
   */
  if(setAltitude) {
     if (!scd30.setAltitudeOffset(altitude)){
        Serial.println("Failed to set altitude offset");
        while(1){ delay(10);}
      }
  }
  Serial.print("Altitude offset: ");
  Serial.print(scd30.getAltitudeOffset());
  Serial.println(" meters");


  /*** Set a temperature offset in hundredths of a degree celcius.
   * Offset value stored in non-volatile memory of SCD30.
   * 1984 offest = 19.84 degrees celcius
   */
  if (setTempOffset) {
      if (!scd30.setTemperatureOffset(tempOffset)){ 
         Serial.println("Failed to set temperature offset");
         while(1){ delay(10);}
      }
   }
  Serial.print("Temperature offset: ");
  Serial.print((float)scd30.getTemperatureOffset()/100.0);
  Serial.println(" degrees C");

  /*** Force the sensor to recalibrate with the given reference value
   * from 400-2000 ppm. Writing a recalibration reference will overwrite
   * any previous self calibration values.
   * Reference value stored in non-volatile memory of SCD30.
   */
   if(setForceRecalibrationReference) {
      if (!scd30.forceRecalibrationWithReference(reference)){
         Serial.println("Failed to force recalibration with reference");
         while(1) { delay(10); }
      }
   }
  Serial.print("Forced Recalibration reference: ");
  Serial.print(scd30.getForcedCalibrationReference());
  Serial.println(" ppm");

  Serial.println("");


  /*** Enable or disable automatic self calibration (ASC).
   * Parameter stored in non-volatile memory of SCD30.
   * Enabling self calibration will override any previously set
   * forced calibration value.
   * ASC needs continuous operation with at least 1 hour
   * 400ppm CO2 concentration daily.
   */
  if(setSelfCalibrationStatus) {
     if (!scd30.selfCalibrationEnabled(enableASC)) {
        Serial.println("Failed to enable or disable Automatic Self Calibration");
        while(1) { delay(10); }
      }
  }
  if (scd30.selfCalibrationEnabled()) {
    Serial.println("Automatic Self Calibration enabled");
  } else {
    Serial.println("Automatic Self Calibration disabled");
  } 

}

void loop() {
  if (scd30.dataReady()) {
    
    if (!scd30.read()){ 
      Serial.println("Error reading sensor data"); 
      return; 
    }

    Serial.print("Temperature: ");
    Serial.print(scd30.temperature);
    Serial.println(" degrees C");

    Serial.print("Temperature offset: ");
    Serial.print((float)scd30.getTemperatureOffset()/100.0);
    Serial.println(" degrees C");

    Serial.print("Altitude offset: ");
    Serial.print(scd30.getAltitudeOffset());
    Serial.println(" meters");

    Serial.print("Ambient pressure offset: ");
    Serial.print(scd30.getAmbientPressureOffset());
    Serial.println(" mBar");
    
    Serial.print("Relative Humidity: ");
    Serial.print(scd30.relative_humidity);
    Serial.println(" %");

    Serial.print("Forced Recalibration reference: ");
    Serial.print(scd30.getForcedCalibrationReference());
    Serial.println(" ppm");

    if (scd30.selfCalibrationEnabled()) {
    Serial.println("Automatic Self Calibration enabled");
    } else {
    Serial.println("Automatic Self Calibration disabled");
    }
    
    Serial.print("CO2: ");
    Serial.print(scd30.CO2, 3);
    Serial.println(" ppm");
    Serial.println("");
  }

  delay(10000);
}
```

**Comment by caternuson** [comment]
I question the actual readability of these values back from the sensor. The datasheet makes it seem like these are write only. I've queried the sensor manufacturer for clarification.

**Comment by scottmil** [comment]
The SCD30 returns the updated ForcedRecalibration reference value when read back immediately after setting the updated value.   However, contrary to the Sensirion documentation,  in my tests using the Adafruit, Seeed-Studio and Sparkfun libraries, the FRC updated reference value apparently does not persist after repowering.

"FRC value is saved in non-volatile memory, the last set FRC value will be used for field-calibration after repowering”

See attached documentation.

**Comment by scottmil** [comment]
There seems to be conflicting information in the SCD30 documentation.   The statement below states that the FRC reference is stored in VOLATILE memory, and the reference 400 is always returned upon repowering:

"The FRC method imposes a permanent update of the CO2 calibration curve which persists after repowering the sensor. The most recently used reference value is retained in volatile memory and can be read out with the command sequence given below. After repowering the sensor, the command will return the standard reference value of 400 ppm. "

The statement above reflects the current behavior of the SCD30.

Whew!

**Comment by caternuson** [comment]
Interesting. Just checked the Interface Description doc version I've been referencing and it's `Version 0.91 – D1 – August 2018`. Looks like there is a newer `Version 1.0 – D1 – May 2020` one available that has that information:
![image](https://user-images.githubusercontent.com/8755041/115771976-6fefac00-a363-11eb-8b79-68d8999ac71f.png)

The older version says this:
![image](https://user-images.githubusercontent.com/8755041/115772246-b7763800-a363-11eb-827d-dc65a17ede00.png)

So looks like maybe an errata in the older documentation?

**Comment by scottmil** [comment]
Carter-

Either errata in the docs, or Sensirion changed their firmware to the newer behavior.

-Scott

**Comment by caternuson** [comment]
Could be. Seems odd to go from persisting to not persisting. FWIW - the one I'm using is also behaving like yours, and per the most recent docs, i.e. - the FRC value does **not** persist. Will be good to hear what Sensirion says about this.

**Comment by scottmil** [comment]
Although Sensirion says the effect of the non-persisted FRC reference adjusts the calibration curve.

**Comment by kevinjwalters** [comment]
I thought this was the other way around??

> `Setting an altitude offset will override any pressure offset.`

**Comment by caternuson** [comment]
@scottmil Please update to the 1.0.7 version of the library when it becomes available. That fixed a minor bug that was potentially messing up the values read back. However, I think the behavior you are seeing with the FRC value is consistent with the latest datasheet. It seems like the actual effects on the readings persist, but you can't actually read the value back after a power cycle.

**Comment by caternuson** [comment]
Closing. Hopefully fixed with 1.0.7.

**Comment by 877dev** [comment]
@scottmil @caternuson have a few questions, if you have time could you possibly help?
https://forums.adafruit.com/viewtopic.php?f=19&t=187307&p=907229#p907170
